### PR TITLE
Return Residual in CalcFactor

### PR DESCRIPTION
**Pull request by Affie**
WIP towards #467

### CODE REVIEW

File: src/NumericalCalculations.jl [hunk 1]
```diff
@@ -195,7 +195,7 @@ function _solveCCWNumeric!( ccwl::Union{CommonConvWrapper{F},
   
   # broadcast updates original view memory location
   ## using CalcFactor legacy path inside (::CalcFactor)
-  _hypoObj = (res, x) -> (target.=x; unrollHypo!(res) )
+  _hypoObj = (x) -> (target.=x; unrollHypo!())
```
Comment (dehann) [inline src/NumericalCalculations.jl:198]: this has to remain `(res, x) ->` as standard for both Optim and NLsolve, and is used as such by both `approxConv` and `approxDeconv`

File: src/ParametricUtils.jl [hunk 2]
```diff
@@ -102,13 +102,10 @@ end
 
 # This is where the actual parametric calculation happens, CalcFactor equivalent for parametric
 function (cfp::CalcFactorMahalanobis)(variables...)
-  #TODO res might be removed see https://github.com/JuliaRobotics/IncrementalInference.jl/issues/467#issuecomment-772987921
-  res = Vector{eltype(variables[1])}(undef, length(cfp.meas))
-  iΣ = cfp.iΣ
-  cfp.calcfactor!(res, cfp.meas, variables...)
-  
+  # call the user function (be careful to call the new CalcFactor version only!!!)
+  res = cfp.calcfactor!(cfp.meas, variables...)
```
Comment (dehann) [inline src/ParametricUtils.jl:106]: Also expand measurements if more than one (`getSample` always returns a `::Tuple`):
```julia
res = cfp.calcfactor!(cfp.meas..., variables...)
```

**Review by dehann — CHANGES_REQUESTED** [review]

File: src/NumericalCalculations.jl (hunk 1 above)
Comment (dehann) [reply src/NumericalCalculations.jl:198]: https://github.com/JuliaRobotics/IncrementalInference.jl/blob/648aa8ca63eaadd69631b1532bc30cfd96efdb68/src/DeconvUtils.jl#L93

Comment (dehann) [reply src/NumericalCalculations.jl:198]: https://github.com/JuliaRobotics/IncrementalInference.jl/blob/648aa8ca63eaadd69631b1532bc30cfd96efdb68/src/NumericalCalculations.jl#L46

Comment (dehann) [reply src/NumericalCalculations.jl:198]: https://github.com/JuliaRobotics/IncrementalInference.jl/blob/648aa8ca63eaadd69631b1532bc30cfd96efdb68/src/NumericalCalculations.jl#L63

File: src/ParametricUtils.jl (hunk 2 above)
Comment (Affie) [reply src/ParametricUtils.jl:106]: This is the measurement from `getParametricMeasurement` it currently returns a `Vector{Float64}`. In what cases would a parametric solve be able to handle more than this? An example or use case would be helpful for development and testing

File: src/NumericalCalculations.jl (hunk 1 above)
Comment (Affie) [reply src/NumericalCalculations.jl:198]: I don't completely follow, It didn't look like the residual was used anywhere, so I just took it out. 
The residual here: 
https://github.com/JuliaRobotics/IncrementalInference.jl/blob/583a1ca2c468775e2e36cb7eb079f17bb8e8ddaa/src/NumericalCalculations.jl#L40
and here:
https://github.com/JuliaRobotics/IncrementalInference.jl/blob/583a1ca2c468775e2e36cb7eb079f17bb8e8ddaa/src/NumericalCalculations.jl#L54
is not populated or used

Comment (dehann) [reply src/NumericalCalculations.jl:198]: I think you might not be looking at the code carefully enough.  To ensure the higher up calls are identical, each Roots and Minimize case has one redundant definition of the `residual`.  Thats why in some cases the parameter `residual` might look excessive.

-------

Look more closely at the Minimize snippet you posted.  The `residual` parameter **is being used** on line 61
https://github.com/JuliaRobotics/IncrementalInference.jl/blob/583a1ca2c468775e2e36cb7eb079f17bb8e8ddaa/src/NumericalCalculations.jl#L54-L61

as I tried to show here: https://github.com/JuliaRobotics/IncrementalInference.jl/pull/1147#discussion_r571094577

Is it necessary to drop the `::AbstractVector{<:Real}` type restriction for autodiff to work?

-----

In the Roots case, the lambda parameter `res` is a **requirement** from NLsolve.jl package, as I tried to show here:
https://github.com/JuliaRobotics/IncrementalInference.jl/pull/1147#discussion_r571093888

File: src/ParametricUtils.jl (hunk 2 above)
Comment (dehann) [reply src/ParametricUtils.jl:106]: okay, lets leave as you have it for now.

File: src/NumericalCalculations.jl (hunk 1 above)
Comment (dehann) [reply src/NumericalCalculations.jl:198]: perhaps we should add the comment:
```julia
# for code symmetry between Minimize (Optim.jl) and Roots (NLsolve.jl), we use standardized lambda `(res, x)-> obj...`
```

File: src/DeconvUtils.jl [hunk 3]
```diff
@@ -90,7 +90,7 @@ function approxDeconv(fcto::DFGFactor,
     #
     
     # lambda with which to find best measurement values
-    hypoObj = (res, tgt) -> (targeti_.=tgt; onehypo!(res) )
+    hypoObj = (tgt) -> (targeti_.=tgt; onehypo!() )
```
Comment (dehann) [inline src/DeconvUtils.jl:93]: i missed this before, same deal: we need standardized lambda interface `(res,x)->obj...`

File: src/NumericalCalculations.jl [hunk 4]
```diff
@@ -43,7 +43,7 @@ function _solveLambdaNumeric( fcttype::Union{F,<:Mixture{N_,F,S,T}},
   #
 
   #
-  r = nlsolve(objResX, u0, inplace=true)
+  r = nlsolve( (res, x) -> res .= objResX(x), u0, inplace=true)
```
Comment (dehann) [inline src/NumericalCalculations.jl:46]: Ah, you are doing the opposite of what I was doing.  I had `residual` explicitly available for in-place operations (which we will still keep somewhere in `CalcFactor.residual::T` for hard in-place factor situations).

I see here now that you are changing to a more implicit usage of `residual` in the Roots (NLsolve.jl) case.  I did not see this before, sorry if I missed it.

File: src/NumericalCalculations.jl (hunk 1 above)
Comment (dehann) [reply src/NumericalCalculations.jl:198]: i think the part I was missing earlier is you added residual in a new lambda at `nlsolve` instead.  I commented separately there a minute ago

**Review by dehann — APPROVED** [review]
https://github.com/JuliaRobotics/IncrementalInference.jl/pull/1147/files#r571239052

File: src/DeconvUtils.jl (hunk 3 above)
Comment (dehann) [reply src/DeconvUtils.jl:93]: ignore

File: src/NumericalCalculations.jl [hunk 5]
```diff
@@ -58,9 +58,9 @@ function _solveLambdaNumeric( fcttype::Union{F,<:Mixture{N_,F,S,T}},
   #
   # wrt #467 allow residual to be standardize for Roots and Minimize and Parametric cases.
   r = if islen1
-    optimize((x) -> (objResX(residual, x); sum(residual.^2)), u0, BFGS() )
+    optimize((x) -> (residual = objResX(x); sum(residual.^2)), u0, BFGS() )
```
Comment (Affie) [inline src/NumericalCalculations.jl:61]: Reading your comments I saw it looks like i forgot the `.=`

Comment (Affie) [reply src/NumericalCalculations.jl:61]: #1150

**Pull request merged**